zephyr/lib/vregion.c

```diff
@@ -0,0 +1,436 @@
+	uint8_t *base;		/* base address of linear allocator */
+	uint8_t *ptr;		/* current alloc pointer */
+	size_t size;		/* size of linear allocator in bytes */
+	size_t used;		/* used bytes in linear allocator */
+	int free_count;		/* number of frees - tuning only */
+};
+
+struct zephyr_heap {
+	struct sys_heap heap;
+	uint8_t *base;		/* base address of zephyr heap allocator */
+	size_t size;		/* size of heap in bytes */
+};
+
+struct vregion {
+	/* region context */
+	uint8_t *base;			/* base address of entire region */
+	size_t size;			/* size of whole region in bytes */
+	size_t pages;			/* size of whole region in pages */
+
+	/* optional text region - RO and Executable */
+	struct vlinear_heap text;	/* text linear heap */
+
+	/* interim heap */
+	struct zephyr_heap interim;	/* interim heap */
+
+	/* interim shared */
+	struct zephyr_heap interim_shared;	/* shared interim heap */
+
+	/* lifetime heap */
+	struct vlinear_heap lifetime;	/* lifetime linear heap */
+
+	/* optional shared static buffer heap */
+	struct vlinear_heap lifetime_shared;	/* shared lifetime linear heap */
+};
+
+/**
+ * @brief Create a new virtual region instance with shared pages.
+ *
+ * Create a new VIRTUAL REGION instance with specified static, dynamic, and shared static sizes.
+ * Total size is the sum of static, dynamic, and shared static sizes.
+ *
+ * @param[in] lifetime_size Size of the virtual region lifetime partition.
+ * @param[in] interim_size Size of the virtual region interim partition.
+ * @param[in] lifetime_shared_size Size of the virtual region shared lifetime partition.
+ * @param[in] interim_shared_size Size of the virtual region shared interim partition.
+ * @param[in] text_size Size of the optional read-only text partition.
+ * @return struct vregion* Pointer to the new virtual region instance, or NULL on failure.
+ */
+struct vregion *vregion_create(size_t lifetime_size, size_t interim_size,
+			       size_t lifetime_shared_size, size_t interim_shared_size,
+			       size_t text_size)
+{
+	struct vregion *vr;
+	uint32_t pages;
+	size_t total_size;
+
+	if (!lifetime_size || !interim_size) {
+		LOG_ERR("error: invalid vregion lifetime size %d or interim size %d",
+			lifetime_size, interim_size);
+		return NULL;
+	}
+
+	/* align up lifetime size and interim size to nearest page */
+	lifetime_size = ALIGN_UP(lifetime_size, CONFIG_MM_DRV_PAGE_SIZE);
+	interim_size = ALIGN_UP(interim_size, CONFIG_MM_DRV_PAGE_SIZE);
+	if (lifetime_shared_size)
+		lifetime_shared_size = ALIGN_UP(lifetime_shared_size, CONFIG_MM_DRV_PAGE_SIZE);
+	if (interim_shared_size)
+		interim_shared_size = ALIGN_UP(interim_shared_size, CONFIG_MM_DRV_PAGE_SIZE);
+	if (text_size)
+		text_size = ALIGN_UP(text_size, CONFIG_MM_DRV_PAGE_SIZE);
+	total_size = lifetime_size + interim_size +
+		lifetime_shared_size + interim_shared_size + text_size;
+
+	/* allocate vregion structure in userspace */
+	vr = rzalloc(SOF_MEM_FLAG_USER, sizeof(*vr));
+	if (!vr)
+		return NULL;
```

Comment:
can we add `sizeof(*vr)` to lifetime allocation and make it the first allocation there? I suppose the user-space thread will have to access it too, so it needs to be in that domain memory